src/cm/communication_manager.rs

```diff
@@ -56,40 +212,146 @@ impl EventChannel {
+            *cm.cm_id.get() = cm_id;
         }
 
-        Ok(CommunicationManager {
-            cm_id: unsafe { NonNull::new(*cm_id.as_mut_ptr()).unwrap_unchecked() },
+        Ok(cm)
+    }
+
+    pub fn get_cm_event(&mut self) -> Result<Event, String> {
+        let mut event_ptr = MaybeUninit::<*mut rdma_cm_event>::uninit();
+
+        let ret = unsafe { rdma_get_cm_event(self.channel.as_ptr(), event_ptr.as_mut_ptr()) };
+
+        if ret < 0 {
+            return Err(format!("Failed to get cm event {:?}", io::Error::last_os_error()));
+        }
+
+        let event = unsafe { NonNull::new(event_ptr.assume_init()).unwrap() };
+
+        let cm_id = unsafe {
+            let raw_cm_id = event.as_ref().id;
+
+            if !raw_cm_id.is_null() {
+                if event.as_ref().event == EventType::ConnectRequest as u32 {
+                    // For connect requests, create a new CommunicationManager
+                    let cm = Arc::new(CommunicationManager {
+                        cm_id: UnsafeCell::new(NonNull::new(raw_cm_id).unwrap()),
+                        user_context: Mutex::new(None),
+                    });
+
+                    let weak_cm = Arc::downgrade(&cm);
+                    let boxed = Box::new(weak_cm);
+                    let raw_box = Box::into_raw(boxed);
+
+                    (*raw_cm_id).context = raw_box as *mut std::ffi::c_void;
+
+                    Some(cm)
+                } else {
+                    // For other events, return the existing CommunicationManager
+                    let context_ptr = (*raw_cm_id).context as *mut Weak<CommunicationManager>;
+                    if !context_ptr.is_null() {
+                        (*context_ptr).clone().upgrade()
+                    } else {
+                        None
+                    }
+                }
+            } else {
+                None
+            }
+        };
+
+        let listener_id = unsafe {
+            let raw_listen_id = event.as_ref().listen_id;
+
+            if !raw_listen_id.is_null() {
+                let context_ptr = (*raw_listen_id).context as *mut Weak<CommunicationManager>;
+                if !context_ptr.is_null() {
+                    (*context_ptr).clone().upgrade()
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        };
+
+        Ok(Event {
+            event,
+            cm_id,
+            listener_id,
         })
     }
+}
+
+unsafe impl Send for EventChannel {}
+unsafe impl Sync for EventChannel {}
 
-    pub fn get_cm_event() -> Result<Event, String> {
-        todo!();
+impl Drop for CommunicationManager {
+    fn drop(&mut self) {
+        let cm_id = unsafe { *self.cm_id.get() };
+        unsafe {
+            let _ = Box::from_raw((*cm_id.as_ptr()).context as *mut Weak<CommunicationManager>);
+            // Clear the context and destroy the ID
+            (*cm_id.as_ptr()).context = std::ptr::null_mut();
```

Comment:
I think we should not clear the `cm_id.context` before `rdma_destroy_id()` returns since this may create a race condition at `get_cm_event()` side.